File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/guild/GuildRequestActor.java
package net.bestia.zoneserver.actor.guild;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import net.bestia.messages.guild.GuildMessage;
import net.bestia.messages.guild.GuildRequestMessage;
import net.bestia.messages.inventory.InventoryListRequestMessage;
import net.bestia.model.dao.GuildDAO;
import net.bestia.zoneserver.actor.SpringExtension;
import net.bestia.zoneserver.actor.zone.ClientMessageActor.RedirectMessage;
import net.bestia.zoneserver.actor.zone.SendClientActor;
import net.bestia.zoneserver.guild.GuildService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.Objects;

/**
 * The actor will reply to guild requests messages which will provide details to
 * the user about the requested guild.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class GuildRequestActor extends AbstractActor {

	public static final String NAME = "requestGuild";

	private final GuildService guildService;
	private final GuildDAO guildDao;
	private final ActorRef sendClient;

	@Autowired
	public GuildRequestActor(GuildService guildService, 
			GuildDAO guildDao) {

		this.guildService = Objects.requireNonNull(guildService);
		this.guildDao = Objects.requireNonNull(guildDao);
		this.sendClient = SpringExtension.actorOf(getContext(), SendClientActor.class);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(GuildRequestMessage.class, this::onRequest)
				.build();
	}
	
	@Override
	public void preStart() throws Exception {
		final RedirectMessage msg = RedirectMessage.get(InventoryListRequestMessage.class);
		context().parent().tell(msg, getSelf());
	}

	private void onRequest(GuildRequestMessage msg) {
		final int guildId = msg.getRequestedGuildId();
		
		// Check if the member is inside the guild.
		if(!guildService.isInGuild(msg.getAccountId(), guildId)) {
			return;
		}
		
		guildDao.findOne(guildId).ifPresent(guild -> {
			final GuildMessage gmsg = new GuildMessage(msg.getAccountId(), guild);
			sendClient.tell(gmsg, getSelf());
		});
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/guild/GuildRequestActor.kt
package net.bestia.zoneserver.actor.guild

import akka.actor.ActorRef
import net.bestia.messages.guild.GuildRequest
import net.bestia.messages.guild.GuildResponse
import net.bestia.model.findOneOrThrow
import net.bestia.model.guild.GuildRepository
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.BQualifier
import net.bestia.zoneserver.actor.routing.DynamicMessageRoutingActor
import net.bestia.zoneserver.guild.GuildService
import org.springframework.beans.factory.annotation.Qualifier

/**
 * The actor will reply to guild requests messages which will provide details to
 * the user about the requested guild.
 *
 * @author Thomas Felix
 */
@Actor
class GuildRequestActor(
    private val guildService: GuildService,
    private val guildDao: GuildRepository,
    @Qualifier(BQualifier.CLIENT_FORWARDER)
    private val sendClientActor: ActorRef
) : DynamicMessageRoutingActor() {

  override fun createReceive(builder: BuilderFacade) {
    builder.matchRedirect(GuildRequest::class.java, this::onRequest)
  }

  private fun onRequest(msg: GuildRequest) {
    val guildId = msg.requestedGuildId

    val isRequesterInGuild = guildService.isInGuild(msg.accountId, guildId)
    if (!isRequesterInGuild) {
      return
    }

    val guild = guildDao.findOneOrThrow(guildId)
    val gmsg = GuildResponse(msg.accountId)
    sendClientActor.tell(gmsg, self)
  }

  companion object {
    const val NAME = "requestGuild"
  }
}
